website/docs/generics.md

```diff
@@ -0,0 +1,1102 @@
+  argument be provided to the class.
+
+```ruby
+class NumericBox
+  extend T::Generic
+  Elem = type_member {{upper: Numeric}}
+end
+
+class IntBox < NumericBox
+  Elem = type_member {{fixed: Integer}}
+end
+
+NumericBox[Integer].new # OK, Integer <: Numeric
+NumericBox[String].new
+#          ^^^^^^ error: `String` is not a subtype of upper bound of `Elem`
+
+IntBox.new
+# ^ Does not need to be invoked like `IntBox[Integer]` because Sorbet can
+#   trivially infer the type argument
+```
+
+Placing the bound on the type member makes it an error to ever instantiate a
+class with that member outside the given bound.
+
+## Generic methods
+
+Methods can also be made generic in Sorbet:
+
+```ruby
+# typed: true
+extend T::Sig
+
+sig do
+  type_parameters(:U)
+    .params(
+      blk: T.proc.returns(T.type_parameter(:U))
+    )
+    .returns(T.type_parameter(:U))
+end
+def with_timer(&blk)
+  start = Time.now
+  res = yield
+  duration = Time.now - start
+  puts "Running block took #{duration.round(1)}s"
+  res
+end
+
+res = with_timer do
+  sleep 2
+  puts 'hello, world!'
+  123
+end
+T.reveal_type(res) # `Integer`
+```
+
+[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Asig%20do%0A%20%20type_parameters%28%3AU%29%0A%20%20%20%20.params%28%0A%20%20%20%20%20%20blk%3A%20T.proc.returns%28T.type_parameter%28%3AU%29%29%0A%20%20%20%20%29%0A%20%20%20%20.returns%28T.type_parameter%28%3AU%29%29%0Aend%0Adef%20with_timer%28%26blk%29%0A%20%20start%20%3D%20Time.now%0A%20%20res%20%3D%20yield%0A%20%20duration%20%3D%20Time.now%20-%20start%0A%20%20puts%20%22Running%20block%20took%20%23%7Bduration.round%281%29%7Ds%22%0A%20%20res%0Aend%0A%0Ares%20%3D%20with_timer%20do%0A%20%20sleep%202%0A%20%20puts%20'hello%2C%20world!'%0A%20%20123%0Aend%0AT.reveal_type%28res%29)
+
+The `type_parameters` method at the top-level of the `sig` block introduces
+generic type variables that can be referenced elsewhere in the signature using
+`T.type_parameter`. Names are specified as Ruby `Symbol` literals. Multiple
+symbol literals can be given to `type_parameters`, like this:
+
+```ruby
+sig do
+  type_parameters(:K, :V)
+    .params(hash: T::Hash[T.type_parameter(:K), T.type_parameter(:V)])
+    .returns([T::Array[T.type_parameter(:K)], T::Array[T.type_parameter(:V)]])
+end
+def keys_and_values(hash)
+  [hash.keys, hash.values]
+end
+```
+
+[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Asig%20do%0A%20%20type_parameters%28%3AK%2C%20%3AV%29%0A%20%20%20%20.params%28hash%3A%20T%3A%3AHash%5BT.type_parameter%28%3AK%29%2C%20T.type_parameter%28%3AV%29%5D%29%0A%20%20%20%20.returns%28%5BT%3A%3AArray%5BT.type_parameter%28%3AK%29%5D%2C%20T%3A%3AArray%5BT.type_parameter%28%3AV%29%5D%5D%29%0Aend%0Adef%20keys_and_values%28hash%29%0A%20%20%5Bhash.keys%2C%20hash.values%5D%0Aend)
+
+Note Sorbet does not support return type inference. This means that doing
```

Comment:
"does not support return type inference" initially made me think that Sorbet did not support ever having generic return values (`.returns(T.type_parameter(:U))`), which is obviously false given examples just a little bit further down.  I think "return type inference" is the usual way to talk about it, but maybe "does not support inferring return types from method bodies" is a better way to say things?

Reply:
Sorry the term is usually "return type polymorphism" not "return type inference"

e.g. https://eli.thegreenplace.net/2018/return-type-polymorphism-in-haskell/

Reply:
"Return type deduction" is what the C++ spec calls it: http://eel.is/c++draft/dcl.spec.auto.general